Lib/zipfile/_path.py

```diff
@@ -257,9 +257,21 @@ def open(self, mode='r', *args, pwd=None, **kwargs):
             if args or kwargs:
                 raise ValueError("encoding args invalid for binary operation")
             return stream
-        else:
-            kwargs["encoding"] = io.text_encoding(kwargs.get("encoding"))
-        return io.TextIOWrapper(stream, *args, **kwargs)
+        # Text mode:
+        encoding = None
+        if args:
+            # Per io.TextIOWrapper, encoding is the first positional arg.
+            # Our API is to pass all *args and **kwargs to TextIOWrapper.
+            # Extract it so we can process it.
+            encoding = args[0]
+            args = args[1:]
+            # We must check this manually as we extract it for processing.
+            if "encoding" in kwargs:
+                raise TypeError(
+                        "argument ('encoding') given by name and position (1)")
+        encoding = kwargs.pop("encoding", encoding)
+        encoding = io.text_encoding(encoding)
```

Comment:
How about instead adding a helper function to perform this work:

```
def _extract_encoding(encoding=None, *args, **kwargs):
    return encoding, args, kwargs
```

And then call it with

```suggestion
        encoding, args, kwargs = _extract_encoding(*args, **kwargs)
        encoding = io.text_encoding(encoding)
```

Let the interpreter do the arg handling.

Happy too to defer refactoring for later.

Reply:
Great suggestion! The code is much cleaner now. EncodingWarning test added.